src/core/search/block_list.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include <absl/types/span.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+#include <optional>
+#include <vector>
+
+#include "core/search/base.h"
+#include "core/search/compressed_sorted_set.h"
+
+namespace dfly::search {
+// BlockList is a container wrapper for CompressedSortedSet / vector<DocId>
+// to divide the full sorted id range into separate blocks. This reduces modification
+// complexity from O(N) to O(logN + K), where K is the max bock size.
+//
+// It tires to balance block sizes in the range [block_size / 2, block_size * 2]
+// by splitting or merging nodes when needed.
+template <typename C /* underlying container */> class BlockList {
+  using BlockIt = typename PMR_NS::vector<C>::iterator;
+  using ConstBlockIt = typename PMR_NS::vector<C>::const_iterator;
+
+ public:
+  BlockList(PMR_NS::memory_resource* mr, size_t block_size = 1000)
+      : block_size_{block_size}, blocks_(mr) {
+  }
+
+  // Insert element, returns true if inserted, false if already present.
+  bool Insert(DocId t);
+
+  // Remove element, returns true if removed, false if not found.
+  bool Remove(DocId t);
+
+  size_t Size() const {
+    return size_;
+  }
+
+  size_t size() const {
+    return size_;
+  }
```

Comment:
why do we need both? and If need both I suggest to implement one via another one

Reply:
Because `Size()` is by our style guide and `size()` is used in templates to be interchangeable with a vector 😁

Reply:
In this case, I would suggest using only size(), because it's really strange